ably/realtime_conn.go

```diff
@@ -86,23 +90,78 @@ func newConn(opts *clientOptions, auth *Auth, callbacks connCallbacks) (*Connect
 		callbacks: callbacks,
 	}
 	c.queue = newMsgQueue(c)
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.watchDisconnect(c.ctx)
 	if !opts.NoConnect {
-		if _, err := c.connect(false); err != nil {
+		if _, err := c.connect(connArgs{}); err != nil {
 			return nil, err
 		}
 	}
 	return c, nil
 }
 
-func (c *Connection) dial(proto string, u *url.URL) (proto.Conn, error) {
+func (c *Connection) dial(proto string, u *url.URL, once bool) (conn proto.Conn, err error) {
+	lg := c.logger().Sugar()
+	start := time.Now()
+	lg.Debugf("Dial protocol=%q url %q ", proto, u.String())
 	// (RTN23b)
 	query := u.Query()
 	query.Add("heartbeats", "true")
 	u.RawQuery = query.Encode()
+	// We first try a single dial to see if we are successful. There is no need to
+	// have this only in the for loop because most cases we will succeed after the
+	// first attempt, let's be prudent here.
+	conn, err = c.dialInternal(proto, u, c.opts.realtimeRequestTimeout())
+	if err != nil {
+		lg.Debugf("Dial Failed in %v with %v", time.Since(start), err)
+		return nil, err
```

Comment:
Should there not be a retry here (or in the caller of `dial`) rather than returning the dial error to the user when the instantiate the client?

Reply:
I had changed the way where retry happen but forgot to update comments. See `Connection.connectWithRetryLoop`  there is no need for `dialInternal` anymore I removed it  in 15287625ff558b4dd8033516056dd9a62f6931c6

Reply:
@gernest sorry I don't think I was clear in what I said.

What I mean is that I don't think this should be returning an error if the initial dial attempt fails with a recoverable error, instead it _should_ be retrying.

To demonstrate, when I run the following code to connect to `192.168.0.200` which I expect to timeout (and after moving [this code](https://github.com/ably/ably-go/blob/feature/conn-open-failures-rtn14/ably/realtime_client.go#L36-L38) to after the error check to avoid a panic):

```go
log.Println("Connecting to 192.168.0.200")

opts := ably.ClientOptions{}.
	Key("xxx:xxx").
	RESTHost("192.168.0.200").
	RealtimeHost("192.168.0.200").
	HTTPOpenTimeout(1 * time.Second)

if _, err := ably.NewRealtime(opts); err != nil {
	log.Fatal(err)
}
```

I get an error after one second and the program exits:

```
2020/10/18 12:23:09 Connecting to 192.168.0.200
2020/10/18 12:23:10 [ErrorInfo :websocket.Dial wss://192.168.0.200:443?echo=true&format=msgpack&heartbeats=true&key=xxx%3Axxx&timestamp=1603020189482: dial tcp 192.168.0.200:443: i/o timeout code=80000 statusCode=0] See https://help.ably.io/error/80000
exit status 1
```

whereas I'd expect that to enter into a retry loop as a dial timeout is a recoverable error.

I think this relates to the design I proposed in https://github.com/ably/ably-go/pull/144#discussion_r404785636 where `NewRealtime` doesn't return an error and has a separate `Start` method that enters an event loop that emits errors to registered handlers (including connection errors), which would look something like:

```go
realtime := ably.NewRealtime(opts)

realtime.Connection.On(ably.CONNECTED, func() {
	// do something now we know we are connected
})

// start enters a blocking event loop, attempts connections
// transparently manages reconnecting / resuming / recovering
if err := realtime.Start(); err != nil {
	// start only returns an error when the connection is unrecoverable
}
// calling realtime.Stop() causes the event loop to gracefully exit
```

cc @tcard, did you consider this design?

Reply:
The spec says that unless `autoConnect == false`, then the connection sequence starts automatically on instantiating the library; how do we incorporate this into this API? `NewRealtime()` shouldn't return an error unless the given `ClientOptions` are invalid.

Reply:
> the connection sequence starts automatically on instantiating the library; how do we incorporate this into this API?

Ah ok I didn't realise that, I proposed starting the loop as a separate step to avoid a race between the library starting to do stuff and the user registering handlers for stuff, although that could be handled by allowing handlers to be registered via the options too, for example:

```go
ably.NewRealtime(
	ably.WithConnectionHandler(func() {
		// handle connection change events
	}),
)
```

Reply:
> although that could be handled by allowing handlers to be registered via the options too

Should we move that discussion to the spec?

---

To reiterate your original point so that it doesn't get lost:

> I don't think this should be returning an error if the initial dial attempt fails with a recoverable error, instead it _should_ be retrying.

This totally should be fixed.

Reply:
> Should we move that discussion to the spec?

I think this is a language issue rather than a spec issue. For example, I don't believe this is a problem in JavaScript as it will not pre-emptively schedule the establishment of connections before the handlers have been registered, so requiring the handlers be registered in the options there doesn't seem necessary.

I may be wrong though, and even if I'm right, it still may be better to make this clear in the spec. @paddybyers WDYT?

> This totally should be fixed.

I agree, but it would be good to agree on how we think it should be fixed, and if we think we should design the API to avoid races between connections being made and handlers being registered or not, as I think now is the correct time to make that design decision (happy to discuss elsewhere though to not block this PR).

Reply:
@lmars 

The spec makes it easier for a callback based API's (JS etc). Go on the other hand looks much of API's aren't callback based.

It is much convenient to keep blocky part around and let users execute them in goroutines to fit the style of the app they are building. `Realtime` is safe for concurrent use.

```
ably.NewRealtime(
	ably.WithConnectionHandler(func() {
		// handle connection change events
	}),
```

Isn't this already covered by event emitter api? we already have `Realtime.Connection.On()` where users can watch for events as they happen.

Reply:
> Isn't this already covered by event emitter api? we already have Realtime.Connection.On() where users can watch for events as they happen.

It is covered by that API yes, but it presents a race condition between the user registering the handler and the library emitting the events.

For example, let's say I'm interested in tracking the connection lifecycle, then under the current design I would write something like:

```go
realtime, err := ably.NewRealtime(...)
if err != nil {
	return err
}
realtime.Connection.On(func(state ably.State) {
	if state == ably.StateConnConnected {
		// handle being connected
	} else {
		// handle not being connected
	}
})
```

But given the `NewRealtime` method has already created the connection, there is no guarantee that my handler is going to receive the `StateConnConnected`, so my program may think it's not connected when it is.

We may of course decide to ignore this race condition, but given we have a choice right now to eliminate it, perhaps we should.

Reply:
Sorry I was unaware of the `autoConnect` option (which I now realise Paddy mentioned above), so we can document the use of that for those that need to avoid the race condition (i.e. instantiate the library with `AutoConnect(false)`, register the handlers, call `realtime.Connect()` to start the connect loop).

@gernest sorry to add that noise to this thread, so the action here is to start the connection loop here in `NewRealtime` and not return dial errors to the user.

Reply:
Hi @lmars , I fixed this in 5c77665eddc7b21c8bc955a6b2ac503e41b84019

add `LogLevel(ably.LogDebug).` in options to see what is happening with the retries.

Reply:
@gernest it looks like the behaviour is now to block in `NewRealtime` until a connection is established, is that the desired behaviour? I thought it would just validate the options, start the event loop and return.

Reply:
@lmars  You can't real do anything without a connection. If you don't care about the connection you can set `autoConnect` to `false` and manually call `Connect`.

This just make sense to me as a sane choice,  however I'm open to new suggestion. You should be aware that even `Connect` will block until a connection is established.

Reply:
> You can't real do anything without a connection

You can, for example, publish messages that will be queued and sent when the connection is established?

I just find it surprising that we're going for an async event loop with events delivered through registered handlers, but starting with a blocking call to `NewRealtime`? As you say, if I care about the connection lifecycle and waiting for the connection to be established (either initially or when the client reconnects), then I can use `autoConnect = false` + register a handler + `client.Connect()`, but if I don't care about the connection lifecycle and just want to instantiate the library and start publishing the messages I want to publish, why do I need to wait?

A non-blocking `NewRealtime` also matches other event-loop based client libraries like [ably-ruby](https://github.com/ably/ably-ruby/blob/v1.1.2/lib/ably/realtime/connection/connection_manager.rb#L30-L33) and [ably-js](https://github.com/ably/ably-js/blob/1.2.3/common/lib/client/connection.js#L38) that also don't block waiting for a connection when instantiating the client (just trigger the start of the connection loop and return).

In summary, I think as long as the behaviour is clearly documented, I'm not too concerned either way, but I was just surprised this was the behaviour when starting an async event loop kind of thing.

Reply:
@lmars @tcard 

> You can, for example, publish messages that will be queued and sent when the connection is established?

Won't this be poor design? I think it makes sense if the `tcp` socket is ready then sending messages even if they are queued (waiting for ably to send `CONNECTED`)  It won't take forever.  If there is no socket available the queue will fill up and waste resources.  In case of reconnection it make sense, because we have historical records of successful obtaining the socket.

Sending messages before even having the socket  that might take even forever to be established should be discouraged. however as you mentioned its fine either way. I mean waiting or not waiting based on the library consumer.  I think we don't need to spin up unnecessary goroutines, library consumers can try to do that at their own discretion based on their app programming models.

> I was just surprised this was the behaviour when starting an async event loop kind of thing.

We seem to differ on what we think about  async aspects of the client. I think in terms of `async` communication. In which we have a `eventLoop` for that. However socket programming is not async in Go so I believe establishing sockets should kinda stay blocky to discourage bad designs.

If we want to spin up goroutines ourselves then we need to rethink some of the `api's` and pass `context.Context` in relevant places for control.

So, in the patch, we are introducing connection opening loop, I think this should not be mistaken with the `eventloop`. Because the later reads from the socket established by the former. Some state events might be emitted with the former is happening but will strictly be one of `CONNECTING`, `DISCONNECTED`, `FAILED`, or `SUSPENDED` but they are happening outside of the main event loop.

what do you think about this?

I'm fine with `NewRealtime` not to block even with `autoConnect=true`. My concerns are

- Should `NewRealtime` also accept `context.Context` for user to be able to terminate the `connectWithRetryLoop` function ?
- Is it okay to encourage users to publish messages with clients that will never be able to open connections  connect (assuming something is wrong)?

Reply:
> > You can, for example, publish messages that will be queued and sent when the connection is established?

> Won't this be poor design? I think it makes sense if the tcp socket is ready then sending messages even if they are queued (waiting for ably to send CONNECTED) It won't take forever.

The spec is explicit that messages are queued whilst the connection is in the `INITIALIZED` state (see [RTL6c](https://www.ably.io/documentation/client-lib-development-guide/features#RTL6c)), so whilst you may consider this to be bad design, I think we have to stick to the spec here, and discuss changes to that design separately.

Also, it looks like this design has already been adopted by ably-go (see [here](https://github.com/ably/ably-go/blob/integration/1.2/ably/realtime_conn.go#L407-L412) where messages are queued when the state is `ConnectionStateInitialized`), so I don't even think we're talking about a change in design here.

> Sending messages before even having the socket that might take even forever to be established should be discouraged

Again, I think we should make a distinction between implementing features to match the 1.2 spec, and discussing potential design flaws in the spec, because, as I understand it, it is expected that messages are queued whilst the connection is not active, regardless of whether that's because it's the first time the connection is being established or the Nth time, since customers should have every expectation that their connection _will_ establish to Ably in a timely manner, and should not be discouraged from having that expectation.

So, I think our opinions differ, but it seems both of us are willing to compromise, so I'll leave it to @tcard to cast his vote and then let's go with his decision (i.e. `NewRealtime` either blocks until initial connection is established, or just starts the connection loop in a goroutine and returns).

Reply:
I don't think it's on the Go library to decide either way. The spec makes it clear that the user-requested operations on the Realtime instance is abstracted from the current state of the connection. That is, attaching, publishing messages, etc. should work (by queueing operations) even if the connection is transiently in a disconnected state. This is a core feature in terms of usability, and pretty much the whole reason we have such complex SDKs.

For this reason, `connect` is only a one-way signal for the library to _start_ the connection loop, same as `close` is a signal for the opposite. It makes sense, given we need to abstract over transient disconnections anyway. As any design decision, as you point out @gernest, it comes with drawbacks, but for users that prefer to avoid them we also offer the `queueMessages` option and the connection event emitter for increased control.

So, while I understand @gernest objections, I think we don't really have an option here but to implement the intent of the spec.

(We also can find a [precedent in the other multithreaded SDK: Java](https://github.com/ably/ably-java/blob/main/lib/src/main/java/io/ably/lib/transport/ConnectionManager.java#L749-L755).)

> Should `NewRealtime` also accept `context.Context` for user to be able to terminate the `connectWithRetryLoop` function ?

That's what `close` is for.

Reply:
> but for users that prefer to avoid them we also offer the `queueMessages` option and the connection event emitter for increased control

I should also add: we don't queue messages indefinitely, but up to the default `connectionStateTtl`. After that, queued and subsequent publishes fail.

Reply:
@lmars @tcard Thanks for your thorough observation and review.
I made the relevant changes which address the points discussed here i.e async connection opening for RealtimeClient

The changes are here 4fb3f431260b7b45b91d0c2f90ef7ea09fdaa1cc PTAL